transport-classes-io_uring/src/main/java/io/netty/channel/uring/AbstractIoUringStreamChannel.java

```diff
@@ -431,18 +436,51 @@ protected void readComplete0(byte op, int res, int flags, short data, int outsta
                     // or convert it to 0 if we could not read because the socket was not readable.
                     allocHandle.lastBytesRead(ioResult("io_uring read", res));
                 } else if (res > 0) {
-                    final int attemptedBytesRead;
                     if (useBufferRing) {
                         short bid = (short) (flags >> Native.IORING_CQE_BUFFER_SHIFT);
                         boolean more = (flags & Native.IORING_CQE_F_BUF_MORE) != 0;
-                        attemptedBytesRead = bufferRing.attemptedBytesRead(bid);
-                        byteBuf = bufferRing.useBuffer(bid, res, more);
+                        if (IoUring.isRecvsendBundleSupported()) {
+                            // If RECVSEND_BUNDLE is supported we need to do a bit more work here.
+                            // In this case we might need to obtain multiple buffers out of the buffer ring as
+                            // multiple of them might have been filled for one recv operation.
+                            // See https://github.com/axboe/liburing/wiki/
+                            // What's-new-with-io_uring-in-6.10#add-support-for-sendrecv-bundles
+                            int read = res;
+                            for (;;) {
+                                int attemptedBytesRead = bufferRing.attemptedBytesRead(bid);
+                                byteBuf = bufferRing.useBuffer(bid, read, more);
+                                read -= byteBuf.readableBytes();
+                                allocHandle.attemptedBytesRead(attemptedBytesRead);
+                                allocHandle.lastBytesRead(byteBuf.readableBytes());
+
+                                assert read >= 0;
+                                if (read == 0) {
+                                    // Just break here, we will handle the byteBuf below.
+                                    break;
+                                }
+                                allocHandle.incMessagesRead(1);
+                                pipeline.fireChannelRead(byteBuf);
+                                byteBuf = null;
+                                if (!allocHandle.continueReading()) {
```

Comment:
I changed this to not use `CompositeByteBuf` and just fire each buffer through the pipeline directly. The reason for this is that otherwise we might hold a lot of memory before the user will be able to do anything with this. This can dramatically increase memory usage as usually the user will do something with the buffer as soon as it is received (which often is releasing it).